Make Go SDK behavior more friendly for server-side use

ian-ross opened the pull request:
### Features and Changes

This PR includes changes based on looking at the Go example in the `examples` repository. I've tried to rework the SDK to match what I would want if I was using it myself for a real application. I don't know if these changes will be something you like, but it's closer to what I would probably be iterating towards if I was developing this SDK for my own use.

What this means in practice:

- The semantics of `Context` and `GrowthBook` have changed. Once `GrowthBook` instance is created from a `Context`, changes to the `Context` are *not* reflected in the `GrowthBook` instance. This decouples the `Context` and `GrowthBook` values. That coupling is something I always found confusing, and it opens up all sorts of potential problems in multi-threaded code. (All Go web applications are multi-threaded by default.)
- There are `With...` update methods on the `GrowthBook` data type to modify any of the values derived from the `Context` after the `GrowthBook` instance is created.
- All of the `With...` methods on `GrowthBook` instances modify the instance in place, **except** for `WithAttributes`, which creates a new `GrowthBook` instance with the relevant attributes that shares all other settings with the original `GrowthBook` instance.

The basic idea here is to set up a `GrowthBook` instance (or instances) during application initialization, then to inject that instance into request handlers, using `WithAttributes` to set up specialized instances in each request handler based on the requesting user.

The rationale for this is the following:

1. When I write web application servers in Go, I (and most other people I know of) inject dependencies into some sort of context that's available in handler functions. (For example, [this](https://echo.labstack.com/docs/context) is what you do if you're using the Echo framework.)
2. The setup of the things that go into the request context usually happens at server startup (things like database connections, as an example).
3. In your handler function, you want to be able to retrieve things from your request context and use them in a way that is thread-safe.

What the new setup looks like in server code:

First, setup in `main`:

```go
	context := growthbook.NewContext().
		WithClientKey("java_NsrWldWd5bxQJZftGsWKl7R2yD2LtAK8C8EUYh9L8").
		WithTrackingCallback(trackingCallback)

	gb = growthbook.New(context)

	gb.LoadFeatures(&growthbook.FeatureRepoOptions{AutoRefresh: true})
```

Then, inject the `GrowthBook` instance into the request handlers somehow. How you do this depends on the framework you're using, but in the example code, I wrote a custom wrapper so I could do it just with the standard library `http` stuff. That looks like this:

```go
// A "wrapped" handler is passed a GrowthBook instance already set up
// with features and attributes based on the incoming request, and
// returns a map to be converted to JSON in the response.
type wrappedFunc func(gb *growthbook.GrowthBook) map[string]any

func wrapHandler(gb *growthbook.GrowthBook, f wrappedFunc) http.HandlerFunc {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Retrieve attributes for user based on request parameters.
		userAttributes := userData(r)

		// Inject the user attributes into the GrowthBook instance and
		// call the wrapped handler function.
		result := f(gb.WithAttributes(userAttributes))

		// Write response: either 200 OK with JSON body, or 400 Bad Request.
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(result)
	})
}
```

The `userData` function is a fake user lookup for the example. In reality, at this point you'd retrieve user claims from some sort of authentication token and look up the user's attributes in a cache and/or database. The end result would still be a set of user attributes.

The critical point here is that the `WithAttributes` method of `GrowthBook` does a thread-safe copy, updating just the attributes. The actual handler function then receives a `GrowthBook` instance with all the user data set up ready for feature evaluation. An actual handler function looks like:

```go
func greetingHandler(gb *growthbook.GrowthBook) map[string]any {
	bannerText := gb.EvalFeature("banner_text").GetValueWithDefault("???")
	return map[string]any{"greeting": bannerText}
}
```

### Testing

All the tests still work with this new setup, and the example server code in `examples/go-example` (see this PR: https://github.com/growthbook/examples/pull/40) shows how to use it (and it works...)

billinghamj commented:
I'd love if there could be a way to support Go `context.Context` making its way through to outbound requests, and the option to specify an HTTP client or roundtripper. Would be very helpful for tracing

ian-ross commented:
> I'd love if there could be a way to support Go `context.Context` making its way through to outbound requests, and the option to specify an HTTP client or transport/roundtripper. Would be very helpful for tracing

Ah yeah, I understand what you mean. (I was confused between `context.Context` and `growthbook.Context` at first!) That's a good idea.

I can add a `*context.Context` to `growthbook.GrowthBook` and default to `context.Background` if it's not set, and that `Context` would be used for HTTP requests. I can do the same sort of thing for the client, using `http.DefaultClient` if it's not set. I think it's easier to require the whole client to be set rather than the roundtripper, just because it gives the user more control.

How does that sound to you? The only tedious interaction I can see would be the case where there is a live streaming request running when the client or context is changed. It's not a big problem, and I need to do some testing on streaming requests today or tomorrow, so I can look at what's needed to make it work seamlessly then.

billinghamj commented:
@ian-ross The key thing with Go context is that it must be included in specific function calls, never captured/persisted on structs etc. So it shouldn't be put on `growthbook.GrowthBook`, assuming you expect that to be reused.

The primary convention is to put it as the first argument on most functions, though of course that would require virtually every method to get a breaking change. It's not best practice, but there's a number of workarounds some libs use to maintain compatibility - e.g. incorporating it into the input params (like Stripe), by having 2 copies of each method (adding "with context" ones), by returning a `Request` struct that can capture the context. In GB's case, you could possibly add a `.WithContext(ctx)` type method on `growthbook.GrowthBook`.

There's also clearly going to be major confusion between the GrowthBook context and the Go context, as in any Go situation, simply saying "context" and using the var name `ctx` would  _always_ mean the Go context. Ideally you might actually rename the GrowthBook one, or give it a name like GBContext to try and disambiguate. Or just accept the breaking change to reduce likelihood of mistakes.

(It's an interface fwiw, so always a pointer, ie `context.Context` not `*context.Context`) 

Just for clarity, do you tend to do a lot of Go work, or more of an occasional thing? Not trying to be negative at all, just want to make sure I'm giving appropriate info etc. Happy to talk through anything if you'd like 🙂 I'm in the GB Slack

We've had quite a lot of experiences using GrowthBook's Go SDK where it behaves in a kinda non-idiomatic way. Originally I was keen to try and support around heavily restructuring the API and getting any needed breaking changes through at once, but it was just an impractical amount of work unfortunately

ian-ross commented:
No negativity inferred... It's useful to hear feedback.

When I wrote the first version of the Go SDK, I was trying to strike a balance between more or less idiomatic Go behavior and keeping things in line with the other GrowthBook SDKs, in particular the JS SDK, which is the main SDK for the platform, and an example of how the GrowthBook people would like their SDKs to look. (They have a great "Build Your Own SDK" page: https://docs.growthbook.io/lib/build-your-own)

I now think that trying to strike that balance might have been a mistake. As you say (in a very kind way!), there are some warts. "Context" is just one of them!

First thing to say is that I don't work for GrowthBook. I'm a contractor, so anything we talk about here will have to be agreed with GrowthBook before I do anything too extensive!

I would be very open to restructuring the Go SDK. One problem is that until last week, I hadn't really *used* it. Then I started updating the Go example application and realized that there were lots of things that weren't ideal. I've made some small-scale changes, but there are other things that could be done.

In terms of things like context and HTTP client, I think it would be easy enough to deal with in an idiomatic way (my description above wasn't very good, and was based on this compromise position). There's really only one place where there are context-dependent operations, and that's in the new feature repository code. It's easy enough to provide a `LoadFeaturesWithContext` to inject a context, and in addition that function can be changed to return a channel that can be used to stop any ongoing background feature fetching goroutines (it's possible to get streaming updates over SSE from the GrowthBook API, and that's handled transparently, but of course "transparently" also means there's no way to interact with it once it's started — as I said, easy enough to provide an affordance there).

I would be genuinely interested to hear a laundry list of complaints about what's not idiomatic and/or what's annoying in the SDK from your perspective. I think we could make a case to @tinahollygb and the other GrowthBook folks that it would look better to have a Go SDK that's really as idiomatic as possible, even if it ends up diverging from the other SDKs in some details.

billinghamj commented:
It might be worth having a user research call to do some feedback and talk through some of the things we've found work well & less well, happy to find a time if you'd like to do that. Will aim to get together some notes on the key things we've found.

Currently we've found it cleanest to instantiate the SDK on each individual use, but that does come with some negatives, and will inherently conflict with the (very exciting!) changes you've been making already to make feature loading work realtime etc.

ian-ross commented:
A call to talk about this stuff sounds good. If you can get a list of issues together, we can discuss them. I'm available all this week (I'm in Europe, so that's easy, but it would be good to get @tinahollygb or @jdorn on the call too and they're in western Canada and the US.)

The feature loading is pretty nice. I've just been testing the SSE updates this morning. (They work!)

And that thing of instantiating everything in each request is something I've also been thinking about a little. I've been trying to organize the data structures and copying semantics for `growthbook.Context` and `growthbook.GrowthBook` to make it more reasonable to set up the `GrowthBook` instance in application initialization and to inject it into some sort of request handling context instead of building a new `GrowthBook` for every request (this sort of thing: https://github.com/ian-ross/growthbook-examples/blob/go-sdk-js-sdk-parity-update/go-example/main.go#L37-L41 and https://github.com/ian-ross/growthbook-examples/blob/go-sdk-js-sdk-parity-update/go-example/handlers.go#L10-L28 — that's just using `net/http`, but you can obviously do the same sort of thing with Echo or whatever framework you use).

That all obviously needs a bit of attention to locking, especially with the autonomous feature loading stuff, but I think it's not too bad. (I was going to say "almost thread-safe", but that's pretty worthless. I believe that it will work without problems in more or less all cases that I can think of, but that doesn't rule out some corner case.)

billinghamj commented:
On the HTTP client - on `growthbook.GrowthBook`, yep that looks great

For `context.Context`, mostly looks good (where it's the first function param)

I think the API on `growthbook.Context` would likely lead to problems though. The general rule is that [structs should never have a context.Context field](https://go.dev/blog/context-and-structs). I think the exception would be clearly-ephemeral structs that are documented as being single-use only. So I'd think having the ability to set contexts on anything long-lived would be a risky move

Though for both HTTP client and `context.Context`, I don't quite understand the usage on `growthbook.Context`. It looks like they might not be used - maybe left over from some different approaches being explored? I would prob suggest just the changes on `growthbook.GrowthBook`

---

If the intended use-case of `growthbook.GrowthBook` is for it to be long-lived, it would seem key to make all the `With` methods immutable, so they return a copy. So e.g. if you're setting the attributes, it won't mutate the attributes on the existing version.

As it currently stands, given the mutability, it would seem risky to do anything but instantiate every time.

We also found that on the tracking callback, we'd still need to have access to the ctx & user ID variables

<img width="1247" alt="Screenshot 2023-07-10 at 18 08 08" src="https://github.com/growthbook/growthbook-golang/assets/905772/72680802-2c11-4af9-ba5b-6c85c08f3d8c">

That usage also inherently means the `(*growthbook.GrowthBook).track` logic is ineffective, as `trackedExperiments` is always empty

---

In terms of `(*growthbook.GrowthBook).track`, the flip-side is that it doesn't have an LRU cache or anything like that. So if you do keep the `growthbook.GrowthBook` instance around and run a few million user IDs through it, it'll endlessly consume more memory. I think at the moment this feature is kinda inherently broken - there's no good way to use it out-of-the-box on the server side.

There are two little improvements that could be made - using a struct as the key, rather than concatenating. And using `struct{}` as the value, rather than `bool`

In the meantime, we're planning on building our own equivalent to `(*growthbook.GrowthBook).track`, possibly in Redis or something. It could possibly make sense for this to be pluggable instead?

billinghamj commented:
Ah also, this is quite a random/under-the-hood thing, and matters less when GB is loading the features for us, but we found that the SDK's logic for parsing JSON was pretty unusual. e.g. in `ParseFeatureMap` and the new `ParseFeatureAPIResponse`

It's better now that it returns `nil` when it's invalid, but typically `json.Unmarshal` directly into the target struct is a great way to go on unmarshaling. For handling more complex types like conditions, Go has options for applying custom unmarshaling logic [1] [2]. That typically makes it much easier and more robust, requiring much less code to deserialize structs, but is capable of handling incredibly complex & polymorphic structures without any trouble

Just one thing that might be nice for the developer experience developing the SDK, but shouldn't really matter from an outside/end-user perspective :)

[1]: https://pkg.go.dev/encoding/json#example-package-CustomMarshalJSON
[2]: https://levelup.gitconnected.com/advanced-json-manipulation-in-go-3a5dfad6008d

ian-ross commented:
> Though for both HTTP client and `context.Context`, I don't quite understand the usage on `growthbook.Context`. It looks like they might not be used - maybe left over from some different approaches being explored? I would prob suggest just the changes on `growthbook.GrowthBook`

Oops. That was just a holdover from an ill-advised earlier set of changes. I'll remove it.

> If the intended use-case of `growthbook.GrowthBook` is for it to be long-lived, it would seem key to make all the `With` methods immutable, so they return a copy. So e.g. if you're setting the attributes, it won't mutate the attributes on the existing version.
> 
> As it currently stands, given the mutability, it would seem risky to do anything but instantiate every time.

Agreed. I started doing that, but the copying is kind of tricky. I tried to do it using a mixture of the `deepcopy` package and custom code, but I think I'm just going to have to write the necessary boilerplate to do it.

> We also found that on the tracking callback, we'd still need to have access to the ctx & user ID variables

This is one of the annoying things about Go's `Context`. In this case, you're building a closure that lives for an indefinite time containing the context, which is just the same as putting it in a struct. I've always thought that Go needs something like Lisp's dynamic variables or React's provider interface for these things — it would be nice to set these things up once and have them propagated down through the call tree.

> <about `(*growthbook.GrowthBook).track`>
> 
> In the meantime, we're planning on building our own equivalent to `(*growthbook.GrowthBook).track`, possibly in Redis or something. It could possibly make sense for this to be pluggable instead?

I think that making it pluggable is a good idea, probably with a better default implementation as well. I made the feature cache pluggable (there's a Redis example that will appear in the docs eventually: https://github.com/ian-ross/growthbook/blob/go-sdk-docs-update/docs/docs/lib/go.mdx#L110-L194), and there's no reason this shouldn't be the same.

ian-ross commented:
> Ah also, this is quite a random/under-the-hood thing, and matters less when GB is loading the features for us, but we found that the SDK's logic for parsing JSON was pretty unusual. e.g. in `ParseFeatureMap` and the new `ParseFeatureAPIResponse`
> 
> It's better now that it returns `nil` when it's invalid, but typically running `json.Unmarshal` directly into the target struct is a great way to go on unmarshaling. For handling more complex types like conditions, Go has options for applying custom unmarshaling logic [1](https://pkg.go.dev/encoding/json#example-package-CustomMarshalJSON) [2](https://levelup.gitconnected.com/advanced-json-manipulation-in-go-3a5dfad6008d). That typically makes it much easier and more robust, requiring much less code to deserialize structs, but is capable of handling incredibly complex & polymorphic structures without any trouble
> 
> Just one thing that might be nice for the developer experience developing the SDK, but shouldn't really matter from an outside/end-user perspective :)

OK, a lot of that is pretty ugly, and I've never really got around to sorting it out properly. I think that what happened was that getting the condition parsing in the first cut of the SDK was a bit of a chore and I did it in this strange way just to get it working. And then followed that bad example for the sake of "consistency"...

Honestly, it's good to get some feedback on this. I kind of cringe a little every time I work on it, but it has been doing the job, so I've not taken the time to improve it.

And I think it *does* matter from the outside perspective. Weird and/or dodgy code doesn't give confidence, does it?

ian-ross commented:
Just to summarise what we've got so far as potential improvements:

1. Clean up the `context.Context` stuff.
2. Get rid of `growthbook.Context` or at least rename it (my choice would probably be to get rid of it and just to have a `With...` builder interface for `growthbook.GrowthBook`; if you really want to have a separate context object, it's just a list of `func (gb *GrowthBook) *GrowthBook` functions, like a lot of packages do for options).
3. Make `growthbook.GrowthBook` truly immutable and safely shareable between goroutines.
4. Make the tracking pluggable and generally just better.
5. Make the JSON serdes less weird.

Does that sound like a good start?

billinghamj commented:
That sounds ideal to me! I guess the only thing you'd need to cover somehow is, if `growthbook.GrowthBook` is immutable, then the feature loading/caching will probably need to exist somewhat "out of band"/global state (like the JS SDK afaik?) or as something pluggable

ian-ross commented:
At the moment, the feature loading operates as a sort of magic shared background thing with some sneaky finalizer stuff to make sure that things get cleaned up properly. I think that could remain in a way that's compatible with immutability of the GrowthBook instance. (Basically, as long as you don't mess with any of the parameters that control feature loading, everything can proceed transparently. If you change the API host or client key or something similar, then any background refresh processes have to stop and restart. I think that could work without any drama, which would be nice.)

I'm going to try to push this forwards a bit in the next couple of days, so there should be something more concrete to discuss later in the week. @tinahollygb had some good suggestions based on her work on the Java and Ruby SDKs that I'm also going to use.

ian-ross commented:
@tinahollygb and @billinghamj — I've made some big changes...

I don't think this covers 100% of James's concerns, but I think it gets relatively close. I've done the following:

1. Replaced `Context` with `Options` and `GrowthBook` with `Client` and changed the behavior of client creation to use the "options" approach, i.e. there aren't so many `With...` methods on `Client` any more.
2. For the smaller number of cases where you really might want to be able to modify client settings (features, overrides, etc.), there *are* `With...` methods (e.g. `WithFeatures`), but they have a completely immutable interface, i.e. the `Client` is cloned, the relevant values are updated in the clone and the clone is returned.
3. There are Go-idiomatic error returns everywhere, and a `context.Context` value is threaded through everywhere that it's relevant.
4. I have changed the way that the SDK works in terms of attributes: these are now passed explicitly to `EvalFeature`, rather than being bound into the client data structure. This fits better with the way that server-based code is likely to work: attributes will probably be derived from authentication information to determine which user is making a request, so the attributes will be bounded by requests, rather than being long-lived (as they are in a front-end application).
5. The logging interface is now much more flexible, and  works well with things like logrus and zerolog.
6. The experiment tracking and feature usage callbacks are now interface-based, so that it's easy to inject your own state-dependent logic into the tracking process. (Actually, I just realised when writing that that a). I've not finished that and b). it might be enough for James even then... The API is right, but the logic isn't yet correct, since it still just uses the same stuff that was there before and there's not yet a way to by-pass the current "call once" logic. I'll look at it some more tomorrow...)
7. I've made all the JSON serialization/deserialization more "normal".

I'll try to sort out the tracking stuff better by Monday so that James can say whether it's enough.

I've also done a little work on the Go example code, including showing how you can use zerolog nicely with the SDK. You can see that code at https://github.com/ian-ross/growthbook-examples/tree/go-sdk-js-sdk-parity-update/go-example

I've also written an example that uses the Echo framework and Logrus for logging. That's interesting because it demonstrates a more realistic approach to injecting client and attributes into response handlers using some custom middleware. That's here: https://github.com/ian-ross/growthbook-examples/tree/go-sdk-js-sdk-parity-update/go-echo-example

I'll try to sort out the tracking tomorrow, but I think this is getting a bit more like a real Go SDK now...

I've not updated any of the docs yet, obviously. I'll do that if you agree that this is a good direction to go.

ian-ross commented:
OK, tracking and caching of tracking is better now. There are interfaces to define trackers for experiments and feature usage, both of which allow for passing Go contexts through along with "extra data", which is good for things like  passing user IDs around. There are also interfaces for caches for experiment and feature usage tracking. The default implementations behave the same as before, but you can do whatever you like with them. The Go Echo example now has a LRU cache for experiment usage, for example: https://github.com/ian-ross/growthbook-examples/blob/go-sdk-js-sdk-parity-update/go-echo-example/tracking.go

@tinahollygb @billinghamj Let me know what you both think. I know that this is quite a large variance from the earlier versions of the Go SDK, but I think that it's a lot better and lot more idiomatic. (James: thanks very much for your feedback, because it was what gave me the impetus and [kind of] the permission to try to make this better. I'm sure it's not quite perfect yet, but I'd be very interested to hear how close it gets to what you need.)

billinghamj commented:
Sorry I haven't checked in - we've been busy and there's a lot of changes here!

I have been taking a little look as I've seen things changing, and generally I think it looks super positive. I suspect we may struggle to feed back effectively without actually giving the new version a try, to see where things fit naturally vs anywhere there could be potential sticking points

Is it at a good point for trying that out? I assume it must be possible for us to point to a specific commit or branch

ian-ross commented:
@billinghamj Yes! I think it's in a good position to use, and it would be *really* helpful to have someone kick it around a bit. You can try all this stuff out on this branch: go-sdk-behavior-update (https://github.com/growthbook/growthbook-golang/tree/go-sdk-behavior-update). And the examples using the new features are here: https://github.com/ian-ross/growthbook-examples/tree/go-sdk-js-sdk-parity-update (in `go-example` and `go-echo-example`). (By the way, which web framework are you using? I've been using Echo for my own projects for a while, but it would be easy enough to modify that example to use some other framework.)

DuffleOne commented:
Hi @ian-ross, @billinghamj and I work together and I've been keeping a keen eye on this PR and the changes. Sorry I haven't chimed in myself but wanted to say hello first and a massive thank you for the time you've taken to address our concerns and feedback! <3

One of my tasks for next week is to implement this against our own branch and run it to get a really good idea about the changes and how they work practically to give you accurate and useful feedback. I don't want to hold you up, is that a good time frame to get hands-on use feedback by/before the end of the next week (4th Aug)?

In regards to the web framework, we use a project from our old company called [crpc](https://github.com/cuvva/cuvva-public-go/tree/master/lib/crpc). We built it ourselves on top of the standard Go HTTP library. I would be over the moon to get in a call with you and run you through what one of our services looks like, how it's initialised, and how we currently use the growthbook SDK, and how we want to use it in the future?

ian-ross commented:
@DuffleOne Hi Laura! More or less any time is good for feedback! The feedback we got from @billinghamj was really useful. Getting detailed user feedback like that doesn't happen too often, and I took as much advantage of it as I could.

We can definitely have a call to talk over how your services work. If you want more detailed information about things, it might be a good idea to have @tinahollygb, @jdorn or someone else from GrowthBook join. I'm an external contractor who just works on the Go SDK, so I can't really say anything about future plans for the SDK, but Tina or Jeremy can. If you just want to pick my brains about these latest changes to the SDK, then we can do that no problem. I'm in Austria, so timezones aren't a problem. We can talk tomorrow or Friday or early next week, depending on whether you want to talk before or after you start work on this. Just let me know when is a good time for you.

I just had a quick look at the example server code for crpc, and it looks like you can do things pretty easily: for example, I would probably just add a `*growthBook.Client` instance to your server structure [here](https://github.com/cuvva/cuvva-public-go/blob/master/lib/crpc/example/server/server.go#L24). Then you'll have access to it in all your handlers. The usual problem that people have is figuring out a way to pass that sort of context through the framework they're using, but in your case, you should be able to make it available right there in the server structure. A `*growthBook.Client` is safe for multi-threaded use (hopefully! it's supposed to be), so you can just share it across the handlers.

DuffleOne commented:
Amazing! 

I'll get on implementing this against our stack in the coming week, and we'll catch up on Friday.

If you have a good idea around crpc, we can also run through it together on Friday as I talk through the integration bits too. I am quite excited to try it all out 😇 

You are absolutely right on the money with your suggest and that is exactly what I hope to do :D

ian-ross commented:
Sounds good! Ping me any time on Friday and we can talk. I'll be out earlier in the morning but I don't have any other fixed appointments the rest of the day that day.

ian-ross commented:
@DuffleOne Just realised I'll be out 12-2 tomorrow as well, so either mid-to-late morning or in the afternoon would work.

DuffleOne commented:
Oh I should have clarified sorry, I'll be busy Friday (tomorrow), I meant from Monday (31st) I'll implement the SDK and get it done quite quickly I hope, but latest feedback to you on the Friday (4th Aug)

ian-ross commented:
No problem! Just ping me when you're ready and we'll set something up.

DuffleOne commented:
@ian-ross Hi Ian! Thanks for your patience. Could we meet today at 14:30? 

I have a Google meet setup here: https://meet.google.com/vav-vzky-zeo

I've implemented the library and would love to run you through it, but I have a few questions on intended best practices, and some thoughts on the use of it

ian-ross commented:
@DuffleOne OK, that works. Talk to you then!

ian-ross commented:
@DuffleOne Sorry, 14:30 UK time, right?

DuffleOne commented:
14:30 UK time :) 

That is 14:30 BST British Summer Time

ian-ross commented:
Got it.

ian-ross commented:
@DuffleOne @billinghamj Thanks for the call! Very useful. To let @tinahollygb and @jdorn know what we talked about:

- There were some questions about type-safety and maybe switching over to using Go generics in some places. They're usable from Go 1.18 so we now have three versions back where we can use them. I would suggest this as something to do after the release of the current changes.
- The Go structured logging package (https://pkg.go.dev/golang.org/x/exp/slog) is now (or soon will be) a part of the standard library, so it makes sense to make our logging interface conform to that. That's an easy change to make, so I'll do it before we release.
- The new experiment tracker interface I wrote needs a little work. This mostly comes down to the fact that the caching behavior that Go applications need often has to work in a multi-server setup, and it's not possible to dictate too much of that to the users of the SDK. The default caching policy isn't appropriate for a multi-server setup, so it would be good to organize things a little differently. I'm going to try to design an API for this that requires users who use the tracking interface to explicitly choose a caching policy, with documentation to try to make it as clear as possible what the consequences of using the default policy are. The aim here is just to prevent people from shooting themselves in the foot.
- One small point that I now realise I didn't write down in my notes clearly enough! @DuffleOne Could you say a few words about the `Experiment` vs. `ExperimentResult` thing you mentioned? It was something about them either both being nil or neither being nil and you having to check both of them...

So, what I'm planning to do:

1. Change the logging interface to be compatible with `slog`.
2. Rework the tracker interface to make it easier to use.
3. Deal with this `Experiment`/`ExperimentResult` thing.
4. Write some documentation.

Once those things are done (probably this weekend), I think we'll be ready to merge and release a v0.2.0. (I would defer any use of generics to a later v0.3.0 release.)

DuffleOne commented:
Thank you for the call @ian-ross!

In regards to the Experiment/ExperimentResult things:

```go
for featureKey := range a.Growthbook.Features() {
	value, err := a.Growthbook.EvalFeatureContext(ctx, featureKey, BuildAttributes(attributeInput), map[string]ksuid.ID{
		"user_id": attributeInput.AuthUser.UserID,
		// TODO: add session ID, not sure how to pass this through yet
	})
	if err != nil {
		return nil, err
	}

	if value.On {
		raw.Values[featureKey] = value.Value
	}

	switch {
	case value.ExperimentResult == nil && value.Experiment == nil:
		// noop

	case value.ExperimentResult != nil && value.Experiment != nil:
		if value.ExperimentResult.InExperiment {
			raw.Experiments[featureKey] = featureconfig.ExperimentResult{
				ExperimentKey: value.Experiment.Key,
				ResultKey:     value.ExperimentResult.Key,
			}
		}

	default:
		return nil, merr.New(ctx, "invalid_feature_value", merr.M{"value": value})
	}
}
```

This is because both are null, or both are set. But we use keys from both structs, so we always want to check they're both set. It is a little esoteric to know that both are set or neither is. And it isn't good practice to assume.

Would be cool to see something more handy in that regard, not sure if a better single structure would work?

ian-ross commented:
Progress:

### Logging

I've updated things to use `slog`, which turns out to be quite amazingly simple. It also makes logging in the Go examples laughably easy (like [this](https://github.com/ian-ross/growthbook-examples/blob/go-sdk-js-sdk-parity-update/go-example/main.go#L17), or [this](https://github.com/ian-ross/growthbook-examples/blob/go-sdk-js-sdk-parity-update/go-echo-example/main.go#L17-L18)).

The only fly in the ointment is that `slog` doesn't work with Go 1.18. Go 1.21 is coming out later this month, so we could just say that the 0.2.0 version of the GrowthBook SDK supports Go 1.19, 1.20 and 1.21. That would be my preferred option, since the `slog` solution really is _much_ better than what I was doing before. That's a question for @tinahollygb or @jdorn though.

### Experiment and ExperimentResult

I've now bundled these up into a single field in the `FeatureResult` type. Either the whole `ExperimentResult` field there is nil or it contains an `ExperimentWithResult` value that has valid experiment and experiment result fields. I think that solves your issue therre, @DuffleOne. I think it's right to say that there really are no cases where the nil-ness of those two values differ, so wrapping them into a single value makes sense.

ian-ross commented:
Also:

### Experiment and feature usage tracking

I have simplified the way this works. Both experiment and feature usage tracking work the same way, so let's just talk about experiment tracking. You supply a value of the `ExperimentTracker` interface in your `Options` when you create a `Client`, and this just has a `Track` method that gets called any time it might be useful to track an experiment. It's up to you what caching behaviour you implement. We provide a default caching implementation to show how this works, which is the existing "never evict" cache. However, I've given it a really long and clear name (`SingleProcessExperimentTrackingCache`) so that no-one will be under any false impressions about it. I will add some clear documentation about how this should work. There's a very simple route to using the simple cache for testing and debugging, so I think that's probably OK: there's enough openness and flexibility for realistic use cases, but there's a quick and easy way to knock together a tracking interface (with or without caching) for testing or development.

@DuffleOne @billinghamj For your use case, I think you should just organize your tracking code as a something that implements the `ExperimentTracker` interface, and you can deal with caching as you like within that. I agree that the separation of the caching logic (i.e. having this `Check`/`Clear` interface as well as the main `Track` interface) wasn't helpful.

Let me know what you think. I'm going to write some documentation now...

ian-ross commented:
@tinahollygb This is now ready for review.

The examples that work with this version are on this branch associated with this PR: https://github.com/growthbook/examples/pull/40 Note that if you want to build those examples using this version of the SDK, you need to clone the `growthbook-golang` repository and the `examples` repositories side-by-side so that the `replace` paths in the examples' `go.mod` files can find the right SDK version. That requirement will obviously go away as soon as we release a v0.2.0.

ian-ross commented:
Documentation updates to go with this are also finished now.

The following 3 PRs all go together:
 - This one.
 - Documentation updates: https://github.com/growthbook/growthbook/pull/1420
 - Updates to examples: https://github.com/growthbook/examples/pull/40

These are all ready for review. The only pre-release changes needed are to fix the examples to use the latest released version of the SDK (they're currently set up to use a pre-release local version of the SDK for demonstration purposes).

@jdorn : I know that @tinahollygb is away for a few weeks, so I'll check in again in a month or so. I'm happy to talk through these changes whenever you like so that we can get them merged.

mvrhov commented:
WOW. This looks so much better than before. We excluded the growthbook because of how clunky the API was before. 👍

ian-ross commented:
@mvrhov Glad you like it! The "clunky" API was also my fault though, just a result of trying too hard to make the GrowthBook Go SDK match up with the way the JS SDK works. Writing a JavaScript API in Go is never going to be pretty. I'm much happier with the way it is now.

billinghamj commented:
Is this one likely to get movement in the near future? :)

ian-ross commented:
I hope so. @tinahollygb was away for a few weeks, but I think she's back now, so maybe we can push it forwards a bit.

I actually had a thought about another potential improvement this morning, based on a recent bug report (#18). I was wondering whether it might be a good idea to split the feature repository management out into a public `FeatureRepository` type that could be explicitly attached to the `Client` instance, instead of relying on the current "magic" way of handling it behind the scenes. The problem with the current setup is that it relies on a finalizer on the `Client` instance to disconnect the feature repository. And finalizers in Go are weirder than I thought they were, which makes that unreliable. I think that having  a `FeatureRepository` that the user can instantiate and connect/disconnect as needed should make it possible get rid of the finalizer magic.

I've not thought through the details yet (just a "dog walk design" idea so far). What do you think?

mkurkov commented:
Hi, I want to push this PR forward, but have some concerns. 

First I don't feel that cloning is the right thing to do. It adds more error prone code, also some parts are confusing - like using JSON marshaling/unmarshaling for cloning. As I understand the main problem is mixing local and global client state. I think we can achieve thread-safety by separating local and global state. So users will first initialize client instance with global(common) configuration and then use this instance providing local parameters. So users will not update common client instance and don't need to have a separate copy. I see Url and Attributes as local state for now. But let me know if I'm missing something.

So normally Go app will have one GrowthBook client instance initialized on start and shared between goroutines, such an instance could be safely updated by separate goroutine
with proper locking, so we will use fresh features rules. But nothing will prevent Go app to create several GB client instances if needed (e.g. with special set of rules for testing).

```
//initialization
options := gb.Options{
  Features: ...,
  APIHost: ...,
  APIKey: ...,
  ...
}
client := gb.NewClient(options)

//usage
var attributes gb.Attributes
json.Unmarshall(`{"user":{"id": 100, "name": "Bob"}, "country": {"name": "USA", "code":"us"}}`, &attributes)
url := "https://example.com"
res := client.EvalFeature("button-color", WithAttributes(attributes), WithUrl(url))
```
This is pretty standard approach in Go libs I think.

Another concern is tracking - I see several tracking callbacks in this PR, not sure I fully understand their meaning.
In original SDK there is only one Tracking Callback that fires when user included in experiment.
I see also that it used for caching results, I think GB has separate feature for this - stickyBucketingService, so it should be provided as this I guess.

Also I'm thinking of refactoring Attributes implementation - to use explicit struct for it instead of dealing with any.

So I plan to clean up code and resolve these issues before merging this. What do you think?

billinghamj commented:
@mkurkov Just picking up on tracking specifically: you mentioned the global vs local state. One issue is that when we're reporting e.g. experiment assignment to our analytics, we need to use the local state + Go ctx as part of that

If you're thinking of a global one that has access to extra data, bear in mind (1) the attributes Growthbook gets are not all we need to do the tracking, and (2) losing typing or having to extract data from a `map[string]any` would be a bit of a pain

On the rest, the work and discussion for this PR happened a long time ago. A lot of effort went into thinking it through and stress-testing the design, so I'd suggest being a bit cautious in what changes. Might be able to get back up to speed and remember all the relevant context, but would not want to do that if this is going to be dropped/forgotten again

billinghamj commented:
Also just to clarify, I assume you've been contracted by Growthbook(?)

mkurkov commented:
Hi @billinghamj, thanks for your reply.
Yes, I'm working on this for Growthbook.
Just to clarify—are you already using code from this branch?

By "global" I mean a common client state, mostly everything we can get from the GB site, and configuration options like logger/HTTP client, etc.
I see that you want both Attributes and ExtraData presented so that we can provide it to the tracking function.
Well, ExtraData is just another part of the local state; I don’t see any problem with it. Or am I missing something?

I envision normal usage as a two-phase process:

-    Initialization of the common client state, starting the update thread (if needed)
-    Calling EvalFeature with the common client state + local state (attributes, extra data, URL)

In the second phase, client users shouldn't be able to update its internal state; only the update thread will do it with proper locking.
So, I don’t see why we need to clone the client state and not just use it in a read-only fashion.

I think we can use something similar to the slog With approach, where it creates a child logger state with some additional meta-info, and, e.g., the HTTP handler can use it instead of
the global logger. It could look like this:

```go

// client.go
type struct RootClient {
  logger Logger
  features Features
  dataSource DataSource
  mu RWMutex
  ...
}

type struct LocalClient {
  root *RootClient
  attributes Attributes
  url string
  extraData any
}

type interface Client {
     EvalFeature(context.Context, string) *Result
}

func (c *RootClient) EvalFeature(ctx context.Context, id string) *Result
func (c *LocalClient) EvalFeature(ctx context.Context, id string) *Result

func (c Client) EvalFeature(id string) *Result
func NewClient(ops ...ClientOption) *RootClient
...

\\ init
client := gb.NewClient(
  WithLogger(log),
  WithDataSource(url, secret),
  WithTrackingCallBack(myTrackingCallback)
)

\\ usage
u := User{name: "John", age: 35, info: "blabla"}
local := client.
  WithUrl("https://mysite.com").    
  WithAttributes(
    RecAttr("user", StrAttr("name", u.name), IntAttr("age",u.age)),
    RecAttr("country", StrAttr("name", "USA"), StrAttr("code", "us"))).
  WithExtraData(u)

buttonColor := local.EvalFeature("button-color").StringValue
newUi := local.EvalFeature("show-new-ui").BoolValue
```

Here, the root client instance is shared among all local clients. Eval could also be called on the root, which would imply that attributes and other local states are empty. WithAttributes and other local state providers update only the local client instance and do not interfere with other instances.

I think the current PR is too large and affects too many areas. However, there are still some outdated sections (such as the usage of deprecated SDK fields) and some new GB features that are not supported.
Some implemented features are now covered by new GB functionalities like StickyBucketService for caching.

So, I have this plan in mind:

1.    Fix the most annoying API issues to make it thread-safe, more usable, and extensible.
2.    Switch the internal representation of attributes and conditions to typed structs instead of using `map[string]any`.
3.    Address the issues that are currently present with the last `cases.json` test suite.
4.    Update the SDK with the latest GB features.

billinghamj commented:
I think your `WithAttributes` option would be a bit painful to use and seems quite unidiomatic - accepting a `map[string]any` would seem to provide better API usability. What would be the benefit of the alternative?

Please make sure context (ctx) params can be passed easily - that is critical to things like the tracking callback. It really should be accepted on anything that could ultimately cause networking/async work - so `EvalFeature` + `NewClient` most likely

I'd suggest supporting generics for the `ExtraData` field. Having to work with an `any` would not be good

mkurkov commented:
> I think your WithAttributes option would be a bit painful to use and seems quite unidiomatic - accepting a map[string]any would seem to provide better API usability. What would be the benefit of the alternative?

This is just an example of internal representation; I plan to use a typed struct for attributes.
We can also provide `WithMapAttributes(map[string]any)` or `WithJSONAttributes(string)`.

By the way, would the process be simplified if `WithAttributes` accepted members of the `AttributesProvider` interface with a `ToGbAttributes()` method? Then you could use a local struct directly if it has a `ToGbAttributes` implementation. Or is it just `map[string]any` most of the time?

> Please make sure context (ctx) params can be passed easily - that is critical to things like the tracking callback. It really should be accepted on anything that could ultimately cause networking/async work - so EvalFeature + NewClient most likely.

I see, yes will provide EvalFeatureContext and NewClientContext variations.

>I'd suggest supporting generics for the ExtraData field. Having to work with an any would not be good

Good idea, but I would postpone it until we resolve the current issues and catch up with the current version of GB in terms of functionality. Moving to generic ExtraData will lead to the "generalization" of other structures and functions,
so it's better to conduct the conversion when the rest of the functionality is ready to use.

billinghamj commented:
I wouldn't expect there to be a lot of use of an interface - in our case we're collecting the data solely for GB's use, not consuming a struct that's used for other purposes. Having several different ways of getting attribute data in might not be super beneficial - possibly could be more error prone and less obvious how to use the lib

I would also encourage considering the context-including methods as being the only ones. Modern day Go pretty much always works with context, and it's easy to provide `context.Background()` if you don't have any use for it

Would also suggest being cautious that any requests made outbound by GB can have tracing added easily (most commonly done by accepting an HTTP client)

abs commented:
Hi folks - @mkurkov just released v0.2.0, which aims to address many/most of the issues raised here - please give it a _Go_ if able!

https://pkg.go.dev/github.com/growthbook/growthbook-golang@v0.2.0

mehdi-sol commented:
@abs Any chance of defining an interface for the logger rather than taking in a `slog.Logger` instance?

mkurkov commented:
Hi @mehdi-sol, I’m glad you liked the new design!

> Any chance of defining an interface for the logger rather than taking in a `slog.Logger` instance, though? Not everyone uses this it and I find having 2 sets of loggers quite awkward to deal with.

I’ve thought about this as well. Currently, Go doesn’t have a common logging interface. That’s why I decided to go with the popular slog library. It’s straightforward to implement the slog.Handle interface and wrap it using slog.New, which allows you to integrate it with your existing logging setup.

Does this approach make sense to you?

mvrhov commented:
This IMHO makes sense as most popular loggers already have such handlers. You've beat me to this.

mehdi-sol commented:
@mkurkov @mvrhov Sorry for the late-ish reply. Was a bit sick. That's completely fair, I'll look into it. But looking more into the source code of the SDK, looks like in a lot of places you guys are not using the context-aware version of the calls, e.g., `.Info` instead of `.InfoContext`. Wondering if this is something you're willing to reconsider -- or maybe have another tip on, because our logging is very dependent on a lot of info carried by the ctx (tracing info being the most important one).

mkurkov commented:
@mehdi-sol, good point. I had initially considered using contexts for tracking calls and managing long-running data source processes. But you are right, we should use context-aware logging methods. I'll fix it in the next release.

mkurkov commented:
@mehdi-sol btw, you can create child Growthbook client with it's own logger parametrized with additional data, like this:
```go
func main() {
  client, err := gb.NewClient(...)
  ...
}

func handleRequest(ctx context.Context, client *gb.Client, r request) {
  traceId := ctx.Value("traceId")
  logger := slog.Default().With("traceId", traceId)
  client,err := client.WithLogger(logger)
  ...
  client,err = client.WithAttributes(gb.Attributes{"user_id": r.user_id})
  ...

  // now all calls to the client will use local logger and attributes
}
```
